src/cargo/core/compiler/context/compilation_files.rs

```diff
@@ -590,7 +590,7 @@ fn compute_metadata(
     // Artifacts compiled for the host should have a different metadata
     // piece than those compiled for the target, so make sure we throw in
     // the unit's `kind` as well
-    unit.kind.hash(&mut hasher);
+    unit.kind.fingerprint_hash().hash(&mut hasher);
```

Comment:
Could we add an explanation here about why use fingerprint hash instead?

BTW you may need to rebase onto master to avoid deprecation warnings from clap v3.2 fails your CI.